### PR TITLE
WIP feature(http): transition towards testable page and action responses

### PR DESCRIPTION
**Pull request by hypeJunction**
Page handlers and action handler can now process direct output rather than relying
on buffer output. This provides a better foundation for testability, as well
improved http responses (including proper http codes).

### CODE REVIEW

**Comment by hypeJunction** [comment]
@mrclay This is the general direction I would like to head with page and action responses. We could then introduce `\Elgg\ActionHandler` and `\Elgg\PageHandler` abstract classes or interfaces that could extended by concreted classes be registered as handlers for `elgg_register_action()` and `elgg_register_page_handler()`

**Comment by hypeJunction** [comment]
We eventually need to eliminate setting headers in views...

**Comment by hypeJunction** [comment]
Or better yet a single `\Elgg\ResourceHandler` for both pages and actions, eventually getting rid of actions as a separate request entry point, relying on request methods (`POST`, `GET`, `PUT`, `DELETE`) to handle requests.

**Comment by hypeJunction** [comment]
Something along these lines:

```php


abstract class ResourceHandler {

	protected $status_code;
	protected $content;

	final public function setStatusCode($status_code = 200);

	final public function getStatusCode();

	final public function setContent($content = '');

	final public function getContent();

	final public function forward($location = '');

	final public function __invoke(array $input = array());

	abstract public function handle(array $input = array(), $resource_view = false);
}

abstract class ActionHandler extends ResourceHandler {

}

elgg_register_resource('blog/add/{container_guid}', 'blog/add'); // default handler that serves a resource view
elgg_register_resource('blog/edit/{guid}', 'blog/edit', \Elgg\Blog\EditResource::class); // custom handler
elgg_register_action('blog/save', \Elgg\Blog\SaveAction::class);

```

File: engine/classes/Elgg/ActionsService.php [hunk 1]
```diff
@@ -97,39 +101,52 @@ public function execute($action, $forwarder = "") {
...
+	
 	/**
 	 * Include an action file with isolated scope
 	 *
 	 * @param string $file File to be interpreted by PHP
 	 * @return void
 	 */
 	protected static function includeFile($file) {
-		include $file;
+		return include $file;
```
Comment (mrclay) [inline engine/classes/Elgg/ActionsService.php]: I'm a little nervous about some 3rd party actions already using return values, but it's probably OK.

Comment (hypeJunction) [inline engine/classes/Elgg/ActionsService.php]: We only look for arrays with specific keys, I think it should be fine.

File: engine/classes/Elgg/Router.php [hunk 2]
```diff
@@ -95,17 +95,24 @@ public function route(\Elgg\Http\Request $request) {
...
-			return true;
+		if ($output) {
+			$ajax_api = _elgg_services()->ajax;
+			if ($ajax_api->isReady()) {
+				$path = implode('/', $request->getUrlSegments());
+				$ajax_api->respondFromOutput($output, "path:$path");
+				return true;
+			}
+
+			return _elgg_services()->response->prepareResponse($output)->send();
```
Comment (hypeJunction) [inline engine/classes/Elgg/Router.php]: This apparently returns the `Response` instance. Will fix.

**Comment by hypeJunction** [comment]
@mrclay If you think we can swallow this for now, I can add tests and docs.

**Comment by mrclay** [comment]
Would it help if forward() threw an exception (#8397)? We could then process all return values in the action service.

**Comment by hypeJunction** [comment]
I really don't like `forward()` and think it needs to die. We abuse it in too many ways, people use it in views, we use it for everything from ajax responses to error pages, and it makes it hard to predict how a certain forward call will behave, we have to use output buffer, and in the end it's not easy to navigate or test.
I would prefer all page and action handlers returned structured data, which can then be used to build  uniform response objects. In the end actions, pages, web services et al are doing the same thing and it would make life a lot easier if we could use the same handler for all of them. I even feel we might have rushed in to using resource views - we seem to have created a new place for madness to happen.

File: engine/classes/Elgg/Http/Response.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+namespace Elgg\Http;
+
+/**
+ * WARNING: API IN FLUX. DO NOT USE DIRECTLY.
+ * 
+ * @since 2.2.0
+ * @access private
+ */
+class Response {
```
Comment (mrclay) [inline engine/classes/Elgg/Http/Response.php]: `ResponseFactory`

File: engine/classes/Elgg/Router.php [hunk 4]
```diff
@@ -95,17 +95,24 @@ public function route(\Elgg\Http\Request $request) {
...
 		}
 
-		echo $output;
-		return $handled || headers_sent();
+		return headers_sent();
```
Comment (mrclay) [inline engine/classes/Elgg/Router.php]: This file's changes basically fixes `elgg/Ajax:path()` and could go in 2.1.

Comment (mrclay) [inline engine/classes/Elgg/Router.php]: Nevermind, it just allows page handlers to return a string. I'm not convinced this is worth it.

Comment (hypeJunction) [inline engine/classes/Elgg/Router.php]: What not worth what?

Comment (hypeJunction) [inline engine/classes/Elgg/Router.php]: I thought of allowing instances of Response to be handled here, but wasn't sure if that's such a great idea.

File: engine/classes/Elgg/ActionsService.php [hunk 5]
```diff
@@ -97,39 +101,52 @@ public function execute($action, $forwarder = "") {
...
 
-		self::includeFile($file);
-		$forward();
-	}
+		if (is_array($result)) {			
+			$content = elgg_extract('content', $result);
+			$status_code = elgg_extract('status_code', $result, 200);
+			$forward_url = elgg_extract('forward_url', $result, REFERRER);
+
+			if (elgg_is_xhr() && $content) {
```
Comment (mrclay) [inline engine/classes/Elgg/ActionsService.php]: If $status_code is in explicitly set to [301, 302, 303, 306, 307, 308] we should honor it even for XHR requests. What you think?

Comment (hypeJunction) [inline engine/classes/Elgg/ActionsService.php]: I suppose we do.

File: engine/classes/Elgg/Router.php [hunk 6]
```diff
@@ -95,17 +95,24 @@ public function route(\Elgg\Http\Request $request) {
 			$handled = call_user_func($function, $segments, $identifier);
 		}
 
-		$output = ob_get_clean();
+		if ($handled === true) {
+			$output = ob_get_clean();
+		} else if (is_string($handled)) {
+			$output = $handled;
+		}
```
Comment (mrclay) [inline engine/classes/Elgg/Router.php]: The effective change here is to allow handlers to return a string. I'd instead allow an array to be returned, and pull out keys like you do in actions.

Comment (hypeJunction) [inline engine/classes/Elgg/Router.php]: Sure. Should we maybe use a convenience function so we don't have to remember all the key names?

File: engine/classes/Elgg/Http/Response.php (hunk 3 above)
Comment (hypeJunction) [inline engine/classes/Elgg/Http/Response.php]: Keep the service name?

**Comment by hypeJunction** [comment]
Replaced by #9779

**Pull request closed without merging**